doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1208 @@
+    
+
+Methods/Slots defined for each DType
+""""""""""""""""""""""""""""""""""""
+
+NEP 41 detailed that all logic should be defined through special methods
+on the DTypes.
+This section will list a specific set of such methods (in the form of 
+Python methods).
+The C-side equivalent slot signature will be summarized below after proposing
+the general C-API for defining new Datatypes.
+Note that while the slots are defined as special Python methods here, this is
+for convenience to the reader and *not* meant to imply the identical exposure
+as a Python API.
+This will need to be proposed in a separate, later, NEP.
+
+Some of the methods may be similar or even reuse existing Python slots.
+User-defined DType classes are discouraged from defining or using Python's
+special slots without feedback, in order to allow defining them later.
+For example ``dtype1 & dtype2`` (or maybe ``|``) could be a shorthand for
+``np.common_dtype``, and comparisons should be defined mainly through casting
+logic.
+
+
+Additional Information
+^^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the more detailed methods below, the following general
+information is currently provided and will be defined on the class:
+
+* ``cls.parametric`` (see also `NEP 40 <NEP40>`_):
+
+  * This flag is mainly to simplify DType and casting creation and
+    allow for performance tweaks.
+  * DTypes which are not parametric must define a canonical dtype instance
+    which should be a singleton.
+  * Parametric dtypes require some additional methods (below).
+  * One can alternatively create a ``ParametricDType`` class for this purpose,
+    although for the C-side API a flag seems simpler.
+
+* ``self.canonical`` method (Alternative: new instance attribute)
+
+  * Instead of byteorder, we may want an ``canonical`` flag (reusing the
+    ISNBO flag – "is native byte order" seems possible here).
+    This flag signals that the data are stored in the default/canonical way.
+    In practice this is always an NBO check, but generalization should be possible.
+    A potential use case is a complex-conjugated instance of Complex which
+    stores ``real - imag`` instead of ``real + imag`` and is thus not
+    the canonical storage.
+
+* ``ensure_canonical(self) -> dtype`` return a new dtype (or ``self``).
+   The returned dtype must have the ``canonical`` flag set.
+
+* ``DType.type`` is the associated scalar type.  ``dtype.type`` will be a
+  class attribute and the current ``dtype.type`` field considered
+  deprecated. This may be relaxed if a usecase arises.
+
+Additionally, existing methods (and C-side fields) will be provided, although
+fields such as the "kind" and and "char" may be set to invalid values on
+the C-side, and access may error on the Python side.
+
+Another example of methods that should be moved to the DType class are the
+various sorting functions, which shall be implemented by defining a method:
+
+* ``dtype_get_sort_function(self, sortkind="stable") -> sortfunction``
```

Comment:
Yes, I should note `compare`, which also exists.

We may want to also just delay this, and instead see if we cannot go the gufunc route to eventually allow fast-path sorting for user-dtypes.  The only problem I see with that is that we need a set of internal gufuncs for all different sort-kinds.

Reply:
True, it would definitely need different `gufunc`, but that code already exists, so should be OK. I think @eric-wieser's adaptations of `gufunc` for `np.linalg` are a nice model to follow.